This file is divided into two parts. In the first one the data frame obtained by the Scraper is cleaned and being prepared for its analysis which is conducted in the second part.

### Part 1 - Cleaning

In [1]:
import pandas as pd

After importing necessary packages we upload the csv file that was created by the class ScraperURL

In [2]:
table = pd.read_csv('C:/Users/jancv/Player_stats.csv')

In [3]:
table.head()

player nationality position           squad     age  \
0         Marko Alvir      hr CRO       MF  Viktoria Plzeň  26-269   
1      Emmanuel Antwi      gh GHA       DF         Příbram  24-253   
2  Oleksandr Azatskyi      ua UKR       DF   Baník Ostrava  27-000   
3         Pavol Bajza      sk SVK       GK        Slovácko  29-131   
4         Jakub Barac      cz CZE       MF  Slovan Liberec  24-162   

   birth_year  games  games_starts minutes  goals  assists  pens_made  \
0        1994      9             0     116      0        0          0   
1        1996      8             2     273      2        0          0   
2        1994      3             0      26      0        0          0   
3        1991      6             6     540      0        0          0   
4        1996      3             1      80      0        0          0   

   pens_att  cards_yellow  cards_red  goals_per90  assists_per90  \
0         0             2          0         0.00            0.0   
1         0             0          0         0.66            0.0   
2         0             0          0         0.00            0.0   
3         0             1          0         0.00            0.0   
4         0             0          0         0.00            0.0   

   goals_assists_per90  goals_pens_per90  goals_assists_pens_per90  
0                 0.00              0.00                      0.00  
1                 0.66              0.66                      0.66  
2                 0.00              0.00                      0.00  
3                 0.00              0.00                      0.00  
4                 0.00              0.00                      0.00

1) Further, we want to find out whether our data frame cointains NaN values. If it is the case, the observation will be dropped.

2) Moreover, we can see that a column 'nationality' needs to be edited in that we are only interested in the 3-letter abbreviation.    The same holds for the age, where the days are not very informative at all, therefore only the age in years will be taken into consideration.

In [4]:
table.isnull().any().any()

False

In [5]:
table = table.dropna()

In [6]:
nationality = []
for nat in table['nationality']:
    nationality.append(nat[3:])
table['nationality'] = nationality

In [7]:
age = []
for ag in table['age']:
    age.append(int(ag[:2]))
table['age'] = age

We found out that at this moment every column where integers or floats are expected do satisfy this condition with the exception of 'minutes'. Thus, we have to convert it from a column of strings to a column of integers.

In [8]:
minutes = []
for minute in table['minutes']:
    minute = minute.replace(",","")
    minutes.append(int(minute))

table['minutes'] = minutes  

In [9]:
table.head()

player nationality position           squad  age  birth_year  \
0         Marko Alvir         CRO       MF  Viktoria Plzeň   26        1994   
1      Emmanuel Antwi         GHA       DF         Příbram   24        1996   
2  Oleksandr Azatskyi         UKR       DF   Baník Ostrava   27        1994   
3         Pavol Bajza         SVK       GK        Slovácko   29        1991   
4         Jakub Barac         CZE       MF  Slovan Liberec   24        1996   

   games  games_starts  minutes  goals  assists  pens_made  pens_att  \
0      9             0      116      0        0          0         0   
1      8             2      273      2        0          0         0   
2      3             0       26      0        0          0         0   
3      6             6      540      0        0          0         0   
4      3             1       80      0        0          0         0   

   cards_yellow  cards_red  goals_per90  assists_per90  goals_assists_per90  \
0             2          0         0.00            0.0                 0.00   
1             0          0         0.66            0.0                 0.66   
2             0          0         0.00            0.0                 0.00   
3             1          0         0.00            0.0                 0.00   
4             0          0         0.00            0.0                 0.00   

   goals_pens_per90  goals_assists_pens_per90  
0              0.00                      0.00  
1              0.66                      0.66  
2              0.00                      0.00  
3              0.00                      0.00  
4              0.00                      0.00

In [10]:
table.to_csv('Final_Player_Table.csv', index = False)

Our data frame is now ready to be analyzed.

### Part 2 - Analysis

In [11]:
import matplotlib.pyplot as plt
import numpy as np

In [12]:
table.head()

player nationality position           squad  age  birth_year  \
0         Marko Alvir         CRO       MF  Viktoria Plzeň   26        1994   
1      Emmanuel Antwi         GHA       DF         Příbram   24        1996   
2  Oleksandr Azatskyi         UKR       DF   Baník Ostrava   27        1994   
3         Pavol Bajza         SVK       GK        Slovácko   29        1991   
4         Jakub Barac         CZE       MF  Slovan Liberec   24        1996   

   games  games_starts  minutes  goals  assists  pens_made  pens_att  \
0      9             0      116      0        0          0         0   
1      8             2      273      2        0          0         0   
2      3             0       26      0        0          0         0   
3      6             6      540      0        0          0         0   
4      3             1       80      0        0          0         0   

   cards_yellow  cards_red  goals_per90  assists_per90  goals_assists_per90  \
0             2          0         0.00            0.0                 0.00   
1             0          0         0.66            0.0                 0.66   
2             0          0         0.00            0.0                 0.00   
3             1          0         0.00            0.0                 0.00   
4             0          0         0.00            0.0                 0.00   

   goals_pens_per90  goals_assists_pens_per90  
0              0.00                      0.00  
1              0.66                      0.66  
2              0.00                      0.00  
3              0.00                      0.00  
4              0.00                      0.00

Our statistics are from the current season 2020/21. There are 18 teams in the Czech Football League which is unsual since normally there are only 16 of them. Due to the coronavirus causing many games to cancel, the relegation part of the last season did not finish, which resulted in the fact that no team was relegated. On the other hand, from the second top league, 2 best clubs were promoted. There have been 14 game days this season, however due to the Covid pandemic, some of them could not be played. Thus, the number of games played by each team varies from 12 to 14. The statistics are therefore biased in this way. 

Firstly, we are interested in nationality.

In [135]:
table.groupby('nationality').count()['player']

nationality
ALB      1
AUT      1
BHR      1
BIH      4
BRA      7
BUL      1
CIV      4
CMR      2
COD      1
COL      1
CRO      4
CUW      1
CZE    342
ESP      3
FRA      6
GAM      1
GEO      1
GER      1
GHA      2
GRE      1
GUI      2
KOR      1
LBR      1
LVA      1
MLI      1
MNE      1
NED      3
NGA      4
NOR      1
POL      2
ROU      2
RUS      4
SEN      1
SRB      6
SVK     34
SWE      1
UKR      3
USA      1
Name: player, dtype: int64

In [136]:
CZ = int(table[table['nationality']=='CZE'].count()['player'])
FO = int(table[table['nationality']!='CZE'].count()['player'])

print(f" There are {CZ} Czech players.")
print(f" There are {FO} foreign players.")
print(f" The ratio of Czech players to all players is {round(CZ/(CZ+FO)*100,1)}%.")

 There are 342 Czech players.
 There are 112 foreign players.
 The ratio of Czech players to all players is 75.3%.


In [40]:
Nat = pd.DataFrame(index=sorted(table['squad'].unique()),columns=['Ratio','Foreigners','Czechs'])

Nat['Czechs'] = table[table['nationality']=='CZE'].groupby('squad').count()['player']
Nat['Foreigners'] = table[table['nationality']!='CZE'].groupby('squad').count()['player']
Nat['Ratio'] = round((Nat['Foreigners']/(Nat['Foreigners']+Nat['Czechs']))*100,1)

Nat_sorted = Nat.sort_values('Ratio', ascending = False)
Nat_sorted

Ratio  Foreigners  Czechs
Karviná          47.8          11      12
České Budĕjov.   43.5          10      13
Fastav Zlín      42.9           9      12
Slovan Liberec   37.9          11      18
Příbram          29.6           8      19
Sparta Prague    28.0           7      18
Slavia Prague    26.1           6      17
Baník Ostrava    24.0           6      19
Opava            21.9           7      25
Viktoria Plzeň   21.7           5      18
Slovácko         17.4           4      19
Sigma Olomouc    17.4           4      19
Zbrojovka Brno   17.2           5      24
FK Pardubice     16.7           4      20
Mladá Boleslav   16.0           4      21
Bohemians 1905   15.4           4      22
Teplice          13.8           4      25
Jablonec         12.5           3      21

In [25]:
print(f"{Nat_sorted.idxmax()['Ratio']} has almost {Nat_sorted.iloc[0,0]} of their squad from abroad. while in the team of {Nat_sorted.idxmin()['Ratio']}, the vast majority of players are Czechs.")

Karviná has almost 47.8 of their squad from abroad. while in the team of Jablonec, the vast majority of players are Czechs.


#### Position

In [133]:
PoS = pd.DataFrame(index=sorted(table['position'].unique()),columns=['Ratio','Foreigners','Czechs'])

PoS['Czechs'] = table[table['nationality']=='CZE'].groupby('position').count()['player']
PoS['Foreigners'] = table[table['nationality']!='CZE'].groupby('position').count()['player']
PoS['Ratio'] = round((PoS['Foreigners']/(PoS['Foreigners']+PoS['Czechs']))*100,1)

Pos_sorted = PoS.sort_values('Ratio',ascending = False)
Pos_sorted

Ratio  Foreigners  Czechs
DF,MF   50.0         6.0       6
FW,MF   47.4         9.0      10
FW      27.5        22.0      58
MF      23.8        43.0     138
DF      22.4        28.0      97
GK      11.1         4.0      32
DF,FW    NaN         NaN       1

Considering only the players that did not change their position, we can see that the highest ratio of foreigners is among forwarders 27.5%. On the contrary, there are 32 Czechs out of 36 among goalkeepers which makes almost 90%.

This notebook is still in progress...

#### Age

In [89]:
AGE = pd.DataFrame(columns = ['all players','more than 6 games'], index = sorted(table['squad'].unique()))
AGE['all players'] = round(table.groupby('squad').mean()['age'].sort_index(),1)
AGE['more than 6 games'] = round(table[table['games']>6].groupby('squad').mean()['age'].sort_index(),1)
AGE.sort_values('all players', ascending = False)

all players  more than 6 games
Viktoria Plzeň         28.0               28.5
Jablonec               27.1               27.6
Sigma Olomouc          26.8               26.8
Slovácko               26.7               27.7
České Budĕjov.         26.5               27.2
Fastav Zlín            26.5               27.2
Baník Ostrava          26.4               27.4
Karviná                26.3               27.1
Teplice                26.1               27.8
Bohemians 1905         25.8               25.9
Slavia Prague          25.5               25.6
Příbram                25.1               26.2
Sparta Prague          25.0               25.8
Zbrojovka Brno         24.8               27.1
Opava                  24.0               27.6
Slovan Liberec         24.0               25.1
FK Pardubice           24.0               25.6
Mladá Boleslav         23.7               24.7

In [46]:
print(f"The average age of all the players is {round(table['age'].mean(),1)} and the mean age of the players which have played more than 6 games is {round(table[table['games']>6]['age'].mean(),1)}.")

The average age of all the players is 25.6 and the mean age of the players which have played more than 6 games is 26.7.


### Cards

In [13]:
table.head()

player nationality position           squad  age  birth_year  \
0         Marko Alvir         CRO       MF  Viktoria Plzeň   26        1994   
1      Emmanuel Antwi         GHA       DF         Příbram   24        1996   
2  Oleksandr Azatskyi         UKR       DF   Baník Ostrava   27        1994   
3         Pavol Bajza         SVK       GK        Slovácko   29        1991   
4         Jakub Barac         CZE       MF  Slovan Liberec   24        1996   

   games  games_starts  minutes  goals  assists  pens_made  pens_att  \
0      9             0      116      0        0          0         0   
1      8             2      273      2        0          0         0   
2      3             0       26      0        0          0         0   
3      6             6      540      0        0          0         0   
4      3             1       80      0        0          0         0   

   cards_yellow  cards_red  goals_per90  assists_per90  goals_assists_per90  \
0             2          0         0.00            0.0                 0.00   
1             0          0         0.66            0.0                 0.66   
2             0          0         0.00            0.0                 0.00   
3             1          0         0.00            0.0                 0.00   
4             0          0         0.00            0.0                 0.00   

   goals_pens_per90  goals_assists_pens_per90  
0              0.00                      0.00  
1              0.66                      0.66  
2              0.00                      0.00  
3              0.00                      0.00  
4              0.00                      0.00

In [33]:
table.groupby('squad').sum().iloc[:,9:11]

cards_yellow  cards_red
squad                                  
Baník Ostrava             24          1
Bohemians 1905            23          1
FK Pardubice              27          1
Fastav Zlín               31          2
Jablonec                  32          0
Karviná                   24          0
Mladá Boleslav            23          0
Opava                     29          1
Příbram                   26          1
Sigma Olomouc             29          0
Slavia Prague             27          0
Slovan Liberec            31          2
Slovácko                  30          3
Sparta Prague             26          2
Teplice                   33          3
Viktoria Plzeň            28          1
Zbrojovka Brno            34          2
České Budĕjov.            27          1

We can see that midfielders tend to have the most yellow cards and defenders are inclined to receive the most red cars.

In [51]:
count_c = table.groupby('position').count().iloc[:,12:14].loc[['DF','FW','GK','MF'],:]
su_c = table.groupby('position').sum().iloc[:,9:11].loc[['DF','FW','GK','MF'],:]
su_c/count_c

cards_yellow  cards_red
position                         
DF            1.128000   0.072000
FW            0.975000   0.050000
GK            0.416667   0.000000
MF            1.182320   0.027624

We can see that midfielders tend to have the most yellow cards and defenders are inclined to receive the most red cars.

In [52]:
count_g = table.groupby('position').count().iloc[:,8:10].loc[['DF','FW','GK','MF'],:]
su_g = table.groupby('position').sum().iloc[:,5:7].loc[['DF','FW','GK','MF'],:]
su_g/count_g

goals   assists
position                    
DF        0.400000  0.392000
FW        1.512500  0.587500
GK        0.027778  0.055556
MF        0.613260  0.546961

Forwards score significantly the most goals but when comparing assists midfielders are only slightly behind them.

In [105]:
print(f"The correlation between games and minutes is {round(np.corrcoef(table['games'],table['minutes'])[1][0],2)}, which is very strong.")


The correlation between games and minutes is 0.87, which is very strong.


In [101]:
print(f"Average time on pitch is {round(table['minutes'].sum()/table['games'].sum(),1)}")

Average time on pitch is 65.4


In [85]:
table.groupby('squad').sum().loc[:,'pens_made']/table.groupby('squad').sum().loc[:,'pens_att']

squad
Baník Ostrava     1.000000
Bohemians 1905    1.000000
FK Pardubice      0.500000
Fastav Zlín       1.000000
Jablonec          1.000000
Karviná           1.000000
Mladá Boleslav    0.000000
Opava             0.333333
Příbram           1.000000
Sigma Olomouc     1.000000
Slavia Prague     1.000000
Slovan Liberec    0.875000
Slovácko          1.000000
Sparta Prague     0.333333
Teplice           0.500000
Viktoria Plzeň    0.666667
Zbrojovka Brno    1.000000
České Budĕjov.    1.000000
dtype: float64

The majority of teams always scored from penalties. Opava and Sparta, however, scored only a third of them.